### PR TITLE
`FnMut` type should be used with a mutable reference

### PR DESCRIPTION
**Pull request by EFanZh**

### CODE REVIEW

**Comment by ehuss** [comment]
Thanks, can you provide some more context on this change? What is the motivation for it?

**Comment by EFanZh** [comment]
Since `C` is of type `FnMut(Oid) -> bool`, calling it would require a mutable reference. Currently, `with_hide_callback` accepts a shared reference (`&C`), intuitively, it should not able to call the function.

File: src/revwalk.rs [hunk 1]
```diff
@@ -170,7 +170,7 @@ impl<'repo> Revwalk<'repo> {
             raw::git_revwalk_add_hide_cb(
                 r.revwalk.raw(),
```
Comment (ynjqqqnt) [inline src/revwalk.rs:171]: 111

File: src/revwalk.rs [hunk 2]
```diff
@@ -170,7 +170,7 @@ impl<'repo> Revwalk<'repo> {
...
                 Some(revwalk_hide_cb::<C>),
```
Comment (ynjqqqnt) [inline src/revwalk.rs:172]: 1
```suggestion
                Some(revwalk_hide_cb::<C>),
```

**Review by ehuss — APPROVED** [review]
OK, I think I understand now. The previous API would allow creating multiple immutable references when it shouldn't. I'm not entirely clear how that can manifest as an issue (since there are several lifetime restrictions in place). It would be interesting to see some example that would be a problem.

I'm also curious if there is some more ergonomic API that could handle this situation. Requiring `&mut` for closures that aren't FnMut seems a bit unfortunate. I tried a few ideas but couldn't come up with anything that would work. If anyone has ideas, it would be appreciated.

I'm going to go ahead and merge since this seems correct to me, and I don't have any ideas on how to make it better.

**Pull request merged**